chain/ethereum/src/network.rs

```diff
@@ -60,11 +68,54 @@ impl EthereumNetworkAdapters {
             })
     }
 
+    pub fn all_cheapest_with_for_call(
+        &self,
+        required_capabilities: &NodeCapabilities,
+    ) -> impl Iterator<Item = Arc<EthereumAdapter>> + '_ {
+        let mut cheapest_sufficient_capability = self
+            .adapters
+            .iter()
+            .filter(|adapter| adapter.call_only())
+            .find(|adapter| &adapter.capabilities >= required_capabilities)
+            .map(|adapter| &adapter.capabilities);
+
+        if cheapest_sufficient_capability.is_none() {
+            cheapest_sufficient_capability = self
+                .adapters
+                .iter()
+                .filter(|adapter| !adapter.call_only())
+                .find(|adapter| &adapter.capabilities >= required_capabilities)
+                .map(|adapter| &adapter.capabilities);
+        }
+
+        self.adapters
+            .iter()
+            .filter(move |adapter| Some(&adapter.capabilities) == cheapest_sufficient_capability)
+            .filter(|adapter| Arc::strong_count(&adapter.adapter) < adapter.limit)
+            .map(|adapter| adapter.adapter.cheap_clone())
+    }
+
+    pub fn cheapest_with_for_call(
+        &self,
+        required_capabilities: &NodeCapabilities,
+    ) -> Result<Arc<EthereumAdapter>, Error> {
+        // Select randomly from the cheapest adapters that have sufficent capabilities.
+        self.all_cheapest_with_for_call(required_capabilities)
+            .choose(&mut rand::thread_rng())
+            .with_context(|| {
+                anyhow!(
+                    "A matching Ethereum network with {:?} was not found.",
+                    required_capabilities
+                )
+            })
+    }
+
     pub fn cheapest(&self) -> Option<Arc<EthereumAdapter>> {
         // EthereumAdapters are sorted by their NodeCapabilities when the EthereumNetworks
         // struct is instantiated so they do not need to be sorted here
         self.adapters
             .iter()
+            .filter(|adapter| !adapter.call_only())
```

Comment:
Given that everywhere we iterate over `adapters` we filter on this, it seems it would be cleaner to separate the lists of ordinary and call-only adapters.